### ISSUE TITLE
spring-boot-gradle-plugin: Do not add findMainClass dependency to each JavaExec task

### ISSUE THREAD

**Issue by pz7** [spring-projects/spring-boot#4639]
Could you please avoid adding `findMainClass` dependency to each `JavaExec` task. Instead add it to spring-boot specific tasks only, like `BootRunTask`.

The reason is when I have my custom Gradle plugin which also uses spring-boot plugin, and where I have my own `JavaExec` task, e.g. for generating additional sources, I don't want my task to depend on spring-boot `findMainClass`, which would create circular dependency, because I want to hook my task as a dependency of `compileJava` task.

Instead of the code in `RunPluginFeatures` (around line 60) which looks like this:

```java
if (task instanceof JavaExec || task instanceof CreateStartScripts) {
    task.dependsOn(FIND_MAIN_CLASS_TASK_NAME);
}
```

I'd rather propose to change it to something like this:
```java
if (task instanceof BootRunTask || task instanceof CreateStartScripts) {
    task.dependsOn(FIND_MAIN_CLASS_TASK_NAME);
}
```

Thanks!

**Comment by philwebb** [comment]
I'm a little nervous that some people might be depending on the existing behavior. I guess they can always add `dependsOn` manually if that's the case but it feels like a risky change for a point release. @wilkinsona any thoughts? Could we do something to allow specific opt-out of the current behavior?

**Comment by wilkinsona** [comment]
TL;DR I think we should change the behaviour in 1.3.1

IMO, the current behaviour could easily be described as a bug. Arguably the plugin is overreaching by configuring the dependency on every JavaExec task. Also, task dependencies can't be removed once they've been added so the current behaviour backs people into a corner. Being more conservative with the dependency configuration will work out of the box for the vast majority of people while also providing a path for people who need a different task dependency configuration.
